softwareComponents/atoms/include/atoms/util.hpp

```diff
@@ -117,11 +117,32 @@ class Defer {
  * Decide if a value fits given type
 */
 template < typename T >
-constexpr bool fitsIn(auto x) {
+constexpr bool fitsIn( auto x ) {
+    using SourceT = decltype( x );
+
     static_assert(std::is_integral_v< T > );
-    static_assert(std::is_integral_v< decltype(x) > );
+    static_assert(std::is_integral_v< SourceT > );
 
     using Trait = std::numeric_limits< T >;
 
-    return x <= Trait::max() && x >= Trait::min();
+    // We have to differentiate three cases based on signedness of the types as
+    // we cannot safely cast values between them
+    if constexpr ( std::is_signed_v<T> == std::is_signed_v< SourceT > ) {
+        return x <= Trait::max() && x >= Trait::min();
+    } else if constexpr ( std::is_unsigned_v<T> && std::is_signed_v< SourceT > ) {
+        if ( x < SourceT( 0 ) )
+            return false;
+
+        if constexpr ( sizeof( T ) >= sizeof( SourceT ) )
+            return true;
+        else
+            return x <= SourceT( Trait::max() );
+    }
+    else { // std::is_signed_v<T> && std::is_unsigned_v< SourceT >
```

Comment:
Tady můžeš dát dokonce static_assert místo komentáře (obecně tomu víc věřím ;) )

Reply:
Chtěl jsem, aby to vizuálně kopírovalo předchozí podmínky (je jasné, že je to poslední možnost, je to tato možnost, ale pomáhá, když se můžeš na ty podmínky přesně podívat).

Reply:
já souhlasím aby to bylo v else větvi (dokonce musí kvůli warningům) - ale místor do komentu bych to dal do static_assertu na další řádek

Reply:
Primárně jsem mířil na vizuální podobnost (a ano, asi bych to ještě zarovnal tak, jak to dělám v této ukázce):
```.cpp
if constexpr        ( std::is_signed_v<T> == std::is_signed_v< SourceT > )    {
	...
} else if constexpr ( std::is_unsigned_v<T> && std::is_signed_v< SourceT > )  {
    ...
} else             /* std::is_signed_v<T> && std::is_unsigned_v< SourceT > */ { 
	...
}
```
vs
```.cpp
if constexpr ( std::is_signed_v<T> == std::is_signed_v< SourceT > ) {
	...
} else if constexpr ( std::is_unsigned_v<T> && std::is_signed_v< SourceT > ) {
    ...
} else { 
    static_assert( std::is_signed_v<T> && std::is_unsigned_v< SourceT > ); 
	...
 }
```

Reply:
Ale tak stejně dobře to můžeš zarovnat na
```cpp
if constexpr        ( std::is_signed_v<T> == std::is_signed_v< SourceT > ) {
    ...
} else if constexpr ( std::is_unsigned_v<T> && std::is_signed_v< SourceT > ) {
    ...
} else { 
    static_assert   ( std::is_signed_v<T> && std::is_unsigned_v< SourceT > ); 
    ...
}
```

Reply:
Přijde mi fajn mít pojištěný od kompilátoru, že to, co tam je napsaný, platí - i když to někdo bude měnit

Reply:
Fair point. Co si myslíš o:
```
if constexpr        ( std::is_signed_v<T> == std::is_signed_v< SourceT > )    {
	...
} else if constexpr ( std::is_unsigned_v<T> && std::is_signed_v< SourceT > )  {
    ...
} else if constexpr ( std::is_signed_v<T> && std::is_unsigned_v< SourceT > ) { 
	...
}
else {
	static_assert( false, "Unreachable")
}
```

Reply:
líbí